src/Compilers/Core/Portable/Emit/EditAndContinue/DeltaMetadataWriter.cs

```diff
@@ -784,6 +810,87 @@ private void PopulateEncLogTableParameters()
             }
         }
 
+        /// <summary>
+        /// CustomAttributes point to their target via the Parent column so we cannot simply output new rows
+        /// in the delta or we would end up with duplicates, but we also don't want to do complex logic to determine
+        /// which attributes have changes, we just emit them all.
+        /// This means our logic for emitting CustomAttributes is to update any existing rows, either from the original
+        /// compilation or subsequent deltas, and only add more if we need to. The EncLog table is the thing that tells
+        /// the runtime which row a CustomAttributes row is (ie, new or existing)
+        /// </summary>
+        private void PopulateEncLogTableCustomAttributes()
+        {
+            var attributeMap = CreateExistingAttributeMap(_previousGeneration.OriginalMetadata.MetadataReader, _previousGeneration.CustomAttributesAdded, out var lastRow);
+
+            var nextCustomAttributeRow = lastRow + 1;
+
+            foreach (var customAttributeTarget in _customAttributeTargets)
+            {
+                int rowId;
+                if (attributeMap.TryGetValue(customAttributeTarget, out var queue) &&
+                    queue.Count > 0)
+                {
+                    // If there is already an attribute for this parent, then pull the next available row number
+                    // off the queue and update it
+                    rowId = queue.Dequeue();
+                }
+                else
+                {
+                    // Otherwise, either this is a new parent that hasn't had an attribute before, or we've run
+                    // out of existing rows to update. Either way, we want to add to the end of the table.
+                    rowId = nextCustomAttributeRow++;
+
+                    // Keep track of the addition for passing to the next generation
+                    _customAttributesAdded[rowId] = customAttributeTarget;
+                }
+
+                _customAttributeRows.Add(rowId);
+                metadata.AddEncLogEntry(
+                    entity: MetadataTokens.CustomAttributeHandle(rowId),
+                    code: EditAndContinueOperation.Default);
+            }
+        }
+
+        /// <summary>
+        /// For each custom attribute target, we need to keep a list of which row numbers are used in the CustomAttribute table
+        /// so that when we emit the new attributes we can overwrite them.
+        /// We also do this for attributes emitted in previous deltas, so the rows may not be a contiguous block
+        /// </summary>
+        private static Dictionary<EntityHandle, Queue<int>> CreateExistingAttributeMap(MetadataReader metadataReader, IReadOnlyDictionary<int, EntityHandle> customAttributesAdded, out int lastRowId)
+        {
+            var result = new Dictionary<EntityHandle, Queue<int>>(HandleComparer.Default);
```

Comment:
>Queue<int> [](http://example.com/codeflow?start=54&length=10)

Since CA table is sorted by Parent, doesn't this Queue always contain a simple range x..y, x < y? If so it'd be more efficient to just remember (x, y) for each parent?

Reply:
Its sorted in the original metadata, and in each generation, but this processes all of them so there are gaps.

Reply:
The CA tables for deltas will be small, but the CA table for baseline can be large. Wonder if it would make sense to optimize for that.

Reply:
> Since CA table is sorted by Parent

So this turns out not to be true... They're grouped by parent, but there is no useful order. Not even all of the attributes for a particular type of parent are grouped together.

Have added a hashset so at least we only create queues that we'll actually use later. One queue per attribute target being emitted in the delta.

Reply:
How come? The spec requires it to be sorted:

![image](https://user-images.githubusercontent.com/41759/120358210-1c09a880-c2bb-11eb-965f-2719f0a65a32.png)

Reply:
https://source.dot.net/#System.Reflection.Metadata/System/Reflection/Metadata/Internal/Tables.cs,813

Reply:
Updated :)